lib/SimpleSAML/Metadata/SAMLBuilder.php

```diff
@@ -716,11 +720,15 @@ private function addCertificate(RoleDescriptor $rd, Configuration $metadata): vo
             if ($key['type'] !== 'X509Certificate') {
                 continue;
             }
+
+            //make sure the key is unique, even when used for multiple entities
+            $keyName = '_' . sha1( $this->entityDescriptor->getEntityID() . $key['keyName']) ?? null;
```

Comment:
@tommy2d I was discussing this with @jaimeperez, but some of the details you've told me this morning went down the drain during my busy workday..  Can you fill them in for us again?

You connect several SPs to your SimpleSAMLphp install and basically have one IDP being TVS.  They require you to deliver a metadata-XML with all metadata of all your SPs, stripped of all their signatures and then globally signed by you, right?  How exactly are they using the KeyName?  Would it still work for you if we change the line above to use sha256 and mix in a secret salt, to avoid collision attacks?

Reply:
I take the metadata of several simpleSAMLphp installs and group them inside a parent document in which they all have to co-exist inside a EntitiesDescriptor element. Metadata-signing is only done on EntitiesDescriptor level; meaning that the individual EntitityDescriptor elements are not signed.

Each of the keys used by any of the EntityDescriptor elements need to have a unique name, even if use the same source-key multiple times. I need to do this because all of the (per-customer) applications that we serve in this way, have their own dedicated simpleSAMLphp install, meaning that if a customer has multiple applications that utilise TVS authentication, there will be multiple instances of the same keyName. This is something that DICTU is now allowing, so i decided to add the entityId to hash source.

What kind of collision attacks are you expecting? For the key name: common practice is using the SHA1 fingerprint stripped of colons. I simply added another factor to make sure I can re-use the same input key multiple times without getting into trouble.

Reply:
We were thinking that a SHA-1 collision might trick the TVS IDP in selecting the wrong key to verify a signature.. That's why I asked if you know how they're using it.. Or are they just enforcing uniqueness?

Reply:
As far as I understand, they just want it to be unique within the EntitiesDescriptor document i supply. As it seems to be common practise to simply use the SHA1 fingerprint for the specifc field, there seems to be no point in trying to prevent a theoretical collision here.

Reply:
I was just looking for the functionality to provide a KeyName, and I found this issue!
I need this to connect to eToegang (specifically: eID from KPN). It requires a KeyName for each KeyInfo in the metadata.

From their documentation:
> Neem voor het certificaat / certificaten binnen de 'SPSSODescriptor' een `<KeyName>`-element op, met een unieke waarde; bij voorkeur de 'thumbprint' van het certificaat

I think that enforcing uniqueness (by hashing with an entityid or secret salt) is not needed, looking at that instruction.

Reply:
For your use case it is indeed not be necessary to perform the additional hashing step. For mine it is, as I will need to include multiple entityDescriptor elements in the same document (LC + DV that utilise the same key and hence wil lead to the same cert name of the fingerprint is to be used directly). Salting is also not needed for my use case as I already pointed out.

Reply:
We've just discussed this a bit more amongst devs and the general consensus is that we can offer a `key_name` setting in the SP-authsource where the name can be configured. We don't want to programmatically calculate some value, because as you see one person needs and extra level of uniqueness and the other wants the thumbprint in there (*).
It only makes sense to configure a static value when you configure a new key.

(*) We should kick KPN in the sack for that, because the thumbprint should never go in the KeyName. The X509Digest-element is meant to hold that.  Identifying a key by it's thumbprint is hazardous.

Reply:
Yes I agree with all of this and I can very wel live with this solution.

Reply:
Yeah, having that option would be enough to be compliant!

Reply:
I'll see what I can do during the weekend, but an addition to this PR is most welcome

Reply:
I'm happy to do it but i have no experience adding configuration settings to the SP auth source.

Reply:
It's quite difficult since we have so many ways of specifying keys..  Including rollover-certs using the `new_` prefix.. I can't promise I have time for this in the near future..